src/page-weekly-snapshot/NYTDataProvider.tsx

```diff
@@ -0,0 +1,171 @@
+import { csvParse, DSVRowString } from "d3";
+import { addDays, isSameDay, startOfToday, subWeeks } from "date-fns";
+import numeral from "numeral";
+import React, { useEffect, useState } from "react";
+
+export type NYTStateRecord = {
+  date: Date;
+  state: string | undefined;
+  cases: number;
+  deaths: number;
+  confirmedCases?: number;
+  confirmedDeaths?: number;
+  probableCases?: number;
+  probableDeaths?: number;
+};
+
+export type NYTCountyRecord = NYTStateRecord & {
+  county: string;
+};
+
+export type NYTData = {
+  state: NYTStateRecord[];
+  counties: NYTCountyRecord[];
+};
+
+type NYTDataMapping = {
+  [stateName: string]: NYTData;
+};
+
+interface NYTDataContext {
+  loading: boolean;
+  error: any;
+  data: NYTDataMapping;
+}
+
+const NYTDataContext = React.createContext<NYTDataContext | undefined>(
+  undefined,
+);
+
+function transformRow(row: DSVRowString): NYTCountyRecord | NYTStateRecord {
+  return {
+    state: row.state,
+    county: row.county,
+    date: row.date ? addDays(new Date(row.date), 1) : startOfToday(),
+    cases: numeral(row["cases"]).value() || 0,
+    confirmedCases: numeral(row["confirmed_cases"]).value() || 0,
+    confirmedDeaths: numeral(row["confirmed_deaths"]).value() || 0,
+    deaths: numeral(row["deaths"]).value() || 0,
+    probableCases: numeral(row["probable_cases"]).value() || 0,
+    probableDeaths: numeral(row["probable_deaths"]).value() || 0,
+  };
+}
+
+export const NYTDataProvider: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const [state, setState] = useState({
+    loading: false,
+    error: null,
+    data: {},
+  });
+
+  async function fetchNYTData() {
+    async function fetchCSVData(url: string) {
+      const response = await fetch(url);
+      const rawCSV = await response.text();
+      return csvParse(rawCSV, transformRow);
+    }
+
+    async function fetchLatestNYTCountyData() {
+      const latestCountyDataURL =
+        "https://raw.githubusercontent.com/nytimes/covid-19-data/master/live/us-counties.csv";
+      return fetchCSVData(latestCountyDataURL);
+    }
+
+    async function fetchHistoricalNYTCountyData() {
+      const historicalCountyDataURL =
+        "https://raw.githubusercontent.com/nytimes/covid-19-data/master/us-counties.csv";
+      return fetchCSVData(historicalCountyDataURL);
+    }
+
+    async function fetchLatestNYTStatesData() {
+      const latestStatesDataURL =
+        "https://raw.githubusercontent.com/nytimes/covid-19-data/master/live/us-states.csv";
+      return fetchCSVData(latestStatesDataURL);
+    }
+
+    async function fetchHistoricalNYTStatesData() {
+      const historicalStatesDataURL =
+        "https://raw.githubusercontent.com/nytimes/covid-19-data/master/us-states.csv";
+      return fetchCSVData(historicalStatesDataURL);
+    }
```

Comment:
seems like you could define these outside the render loop since they don't depend on any local variables